File: frontend/src/types/pagination.ts
export interface Page<T> {
  items: T[];
  page: number;
  perPage: number;
  total: number;
  totalPages: number;
}

export interface PageRequest {
  page?: number;
  perPage?: number;
}

export const DEFAULT_PER_PAGE = 20;
export const PER_PAGE_OPTIONS = [10, 20, 50, 100] as const;

File: backend/modules/restaurant/schemas/pagination.py
"""Shared pagination response models."""

from typing import Generic, TypeVar

from pydantic import BaseModel, Field


ItemT = TypeVar("ItemT")


class PaginatedResponse(BaseModel, Generic[ItemT]):
    """Typed page envelope used by collection endpoints."""

    items: list[ItemT] = Field(default_factory=list)
    page: int = Field(ge=1)
    per_page: int = Field(ge=1, le=100)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


def total_pages(total: int, per_page: int) -> int:
    """Return the number of pages required for a collection."""

    return (total + per_page - 1) // per_page if total else 0
